### PR TITLE
[FEATURE] Create PostGISExecuteAndLoadSQL.py

### PR DESCRIPTION
**Pull request by anitagraser**
## Description
This script closes a gap in the existing Processing tools for PostGIS by enabling the user to load the results of the query as a layer.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] ~~Commits which fix bugs include `fixes #11111` in the commit message next to the description~~
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] ~~Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented~~
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] ~~New unit tests have been added for core changes~~
- [x] I have run autopep8 manually ~~[the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit~~

### CODE REVIEW

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    def processAlgorithm(self, parameters, context, feedback):
+        connection = self.parameterAsString(parameters, self.DATABASE, context)
+        #db = postgis.GeoDB.from_name(connection)
+        uri = postgis.uri_from_name(connection)
+        sql = self.parameterAsString(parameters, self.SQL, context).replace('\n', ' ')
+        QgsApplication.messageLog().logMessage(str(sql), level=Qgis.Info)
+        uri.setDataSource("","("+sql+")", "geom", "","id")
+        vlayer = QgsVectorLayer(uri.uri(), "layername", "postgres")
+        QgsApplication.messageLog().logMessage("Valid layer: "+str(vlayer.isValid()), level=Qgis.Info)
+        #QgsProject.instance().addMapLayer(vlayer)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Was this the original intention here? If so, you could do this by:

    def processAlgorithm(self, parameters, context, feedback):
         # ...everything required to build up the uri...
         layer = QgsProcessingUtils.mapLayerFromString( uri.uri(), context )
         if layer is None:
              raise QgsProcessingException(self.tr("some helpful error message"))

         context.addLayerToLoadOnCompletion( layer.id(), QgsProcessingContext.LayerDetails( 'some nice name', context.project(), self.OUTPUT ) )

         return { self.OUTPUT: layer.id() }
        
If you go this way, you'd need to change OUTPUT from a QgsProcessingParameterFeatureSink to a QgsProcessingOutputVectorLayer output.

Comment (anitagraser) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Yes, that was the original intention. Unfortunately, I'm having no success with QgsProcessingUtils.mapLayerFromString( uri.uri(), context ). The resulting layer is always None even though the uri works with the previous code.

Comment (anitagraser) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Not sure how to debug mapLayerFromString's silent refusal to return a layer.

Comment (anitagraser) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Thanks, that worked!

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return 'postgisexecuteandloadsql'
+
+    def displayName(self):
+        return self.tr('PostGIS execute and load SQL')
+
+    def processAlgorithm(self, parameters, context, feedback):
+        connection = self.parameterAsString(parameters, self.DATABASE, context)
+        uri = postgis.uri_from_name(connection)
+        sql = self.parameterAsString(parameters, self.SQL, context).replace('\n', ' ')
+        QgsApplication.messageLog().logMessage(str(sql), level=Qgis.Info)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Drop this - or use feedback.pushInfo instead

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        uri.setDataSource("","("+sql+")", "geom", "","id")
+        
+        vlayer = QgsVectorLayer(uri.uri(), "layername", "postgres")
+        
+        if vlayer is None:
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: I don't think this is strictly necessary - vlayer will never be None, so you could just skip this and leave the isValid() check only

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            raise QgsProcessingException(self.tr("Got None instead of vector layer object."))
+            return {}
+        else:
+            QgsApplication.messageLog().logMessage("Valid layer: "+str(vlayer.isValid()), level=Qgis.Info)
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: drop or use feedback.pushInfo

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+            if not vlayer.isValid():
+                raise QgsProcessingException(self.tr("This layer is invalid! Please check the PostGIS log for error messages."))
+                return {}
```
Comment (nyalldawson) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Drop

**Review by nyalldawson — COMMENTED** [review]
Looking good - don't forget to run the prepare-commit script (or autopep8 manually) too

**Comment by nyalldawson** [comment]
Oh, I just realised you're missing the required changes to QgisAlgorithmProvider.py too

**Comment by anitagraser** [comment]
Thanks! QgisAlgorithmProvider.py wasn't on my radar.

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    def displayName(self):
+        return self.tr('PostGIS execute and load SQL')
+
+    def processAlgorithm(self, parameters, context, feedback):
+        connection = self.parameterAsString(parameters, self.DATABASE, context)
+        uri = postgis.uri_from_name(connection)
+        sql = self.parameterAsString(parameters, self.SQL, context)
+        sql = sql.replace('\n', ' ')
+        uri.setDataSource("", "("+sql+")", "geom", "", "id")
```
Comment (Gustry) [inline python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: Maybe in the help or in the parameter description, you should add some notes saying that columns **must** must be called "geom" and "id"?

Comment (anitagraser) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: The SQL query description is "SQL query (must return unique id and geom field)"

**Comment by anitagraser** [comment]
I see my pep8 efforts were insufficient, will try autopep tomorrow

File: python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py (hunk 6 above)
Comment (nyalldawson) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: The "import to postgis" algorithm gives users a string input to change these column names - maybe that would be a good move here. The "geom column" name parameter could even be optional to allow geometryless result tables...

Comment (anitagraser) [reply python/plugins/processing/algs/qgis/PostGISExecuteAndLoadSQL.py]: I'm not quite sure how to create the layer if the geom column value is empty. Everything else should be in place now.

**Comment by m-kuhn** [comment]
Is there anything missing to merge this?

**Comment by anitagraser** [comment]
I think it should be fine

**Review by Gustry — COMMENTED** [review]
Thanks for this feature!
Just some small comments.
Maybe also provide a basic help text? (we should make it for each algs IMHO)

**Comment by DelazJ** [comment]
>  Maybe also provide a basic help text? (we should make it for each algs IMHO)

Yes, please! This is more than appreciated when it comes to document the alg (even more when the writer's working environment/data does not necessarily fit the alg's).
And more generally, I think the end user is less lost when there's a kind of description than nothing.

**Comment by anitagraser** [comment]
Do you mean implementing the shortHelpString function or writing something in the documentation?

**Comment by Gustry** [comment]
Hum, I'm wondering between `shortHelpString` and https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/help/qgis.yaml

**Comment by nyalldawson** [comment]
Please don't merge yet - there was an additional test I wanted to conduct here first

**Comment by anitagraser** [comment]
None of the other QGIS algs seems to have a shortHelpString

**Comment by anitagraser** [comment]
Ok, added to https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/help/qgis.yaml

**Comment by m-kuhn** [comment]
Oops, sorry, missed this one comment.
4 months to fix it ;)

**Pull request merged**